server/src/main/java/org/elasticsearch/indices/recovery/MultiFileTransfer.java

```diff
@@ -0,0 +1,211 @@
+ * To ensure the recover target never buffers more than {@code maxConcurrentFileChunks} file chunks, we allow the sender/requestor to send
+ * only up to {@code maxConcurrentFileChunks} file chunk requests from the last flushed (and acknowledged) file chunk. We leverage the local
+ * checkpoint tracker for this purpose. We generate a new sequence number and assign it to each file chunk request before sending; then mark
+ * that sequence number as processed when we receive a response for the corresponding file chunk request. With the local checkpoint tracker,
+ * we know the last acknowledged-flushed file-chunk is a file chunk whose {@code requestSeqId} equals to the local checkpoint because the
+ * recover target can flush all file chunks up to the local checkpoint.
+ * <p>
+ * When the number of un-replied file chunk requests reaches the limit (i.e. the gap between the max_seq_no and the local checkpoint is
+ * greater than {@code maxConcurrentFileChunks}), the sending/requesting thread will abort its execution. That process will be resumed by
+ * one of the networking threads which receive/handle the responses of the current pending file chunk requests. This process will continue
+ * until all chunk requests are sent/responded.
+ */
+abstract class MultiFileTransfer<Request extends MultiFileTransfer.ChunkRequest> implements Closeable {
+    private Status status = Status.PROCESSING;
+    private final Logger logger;
+    private final ActionListener<Void> listener;
+    private final LocalCheckpointTracker requestSeqIdTracker = new LocalCheckpointTracker(NO_OPS_PERFORMED, NO_OPS_PERFORMED);
+    private final AsyncIOProcessor<FileChunkResponseItem> processor;
+    private final int maxConcurrentFileChunks;
+    private StoreFileMetaData currentFile = null;
+    private final Iterator<StoreFileMetaData> remainingFiles;
+    private Tuple<StoreFileMetaData, Request> readAheadRequest = null;
+
+    protected MultiFileTransfer(Logger logger, ThreadContext threadContext, ActionListener<Void> listener,
+                                int maxConcurrentFileChunks, List<StoreFileMetaData> files) {
+        this.logger = logger;
+        this.maxConcurrentFileChunks = maxConcurrentFileChunks;
+        this.listener = listener;
+        this.processor = new AsyncIOProcessor<>(logger, maxConcurrentFileChunks, threadContext) {
+            @Override
+            protected void write(List<Tuple<FileChunkResponseItem, Consumer<Exception>>> items) {
+                handleItems(items);
+            }
+        };
+        this.remainingFiles = files.iterator();
+    }
+
+    public final void start() {
+        addItem(UNASSIGNED_SEQ_NO, null, null); // put a dummy item to start the processor
+    }
+
+    private void addItem(long requestSeqId, StoreFileMetaData md, Exception failure) {
+        processor.put(new FileChunkResponseItem(requestSeqId, md, failure), e -> { assert e == null : e; });
```

Comment:
I wonder if we should somehow assert that this never blocks? (since `AsyncIOProcessor` might be using blocking `ArrayBlockingQueue#put` on the item). It seems to me logically that's not an option with the current code, but maybe something we should make sure of to avoid (admittedly super unlikely) deadlocks in the future?

Reply:
The current implementation is perfectly fine as we control the number of possible items in the queue. We can assert the remaining capacity before putting an item but this option subjects to a race condition. Another option is to use "offer" instead of "put" in AsyncIOProcessor if the assertion is enabled and blocking is not allowed. I will think about it more.

Reply:
Yea, I didn't have a good idea on how to do this now either. Just figured I'd raise it :) Either way, once we start using this in tests that use `DeterministicTaskQueue` we'll probably be guarded against deadlocks/blocking anyway :)